book/src/entry-tree.md

```diff
@@ -90,6 +90,41 @@ preserving the chain of origination.
 (i.e. dealing with forks) will have to be used for the most recent entries in
 the EntryTree.
 
+### EntryTree Design
+
+1. Entries in the EntryTree are stored as key-value pairs, where the key is the concatenated
+slot index and blob index for an entry, and the value is the entry data. Note blob indexes are zero-based for each slot (i.e. they're slot-relative).
+
+2. The EntryTree maintains metadata for each slot, containing:
+      a. `slot_index` - The index of this slot
+      b. `num_blocks` - The number of blocks in the slot (used for chaining to a previous slot)
+      c. `consumed` - The highest blob index `n`, such that for all `m < n`, there exists a blob in this slot with blob index equal to `n` (i.e. the highest consecutive blob index).
+      d. `received` - The highest received blob index for the slot
+      e. `next_slots` - A list of future slots this slot could chain to. Used when rebuilding
+      the ledger to find possible fork points.
+      f. `previous_slot` - The previous slot index that this slot chains to, used to identify
+      whether the entries in this slot are of interest to any subscriptions
+      g. `highest_tick` - Tick height of the highest received blob (used to identify when a slot is full)
+
+3. Chaining - When a blob for a new slot `x` arrives, we check the number of blocks (`num_blocks`) for that new slot (this information is encoded in the blob). We then know that this new slot chains to slot `x - num_blocks`.
+
+4. Subscriptions - The EntryTree records a set of slots that have been "subscribed" to. This means entries that chain to these slots will be sent on the EntryTree channel for consumption by the ReplayStage. See the `EntryTree APIs` for details.
+
+### EntryTree APIs
+
+The EntryTree offers a subscription based API that ReplayStage uses to ask for entries it's interested in. The entries will be sent on a channel exposed by the EntryTree. These subscription API's are as follows:
+   1. `fn subscribe(slot_index: u64) -> ()`: Let `s` be the slot such that `s.slot_index == slot_index`. This `subscribe(slot_index)` function call, tells the EntryTree to return all entries `e` such that:
```

Comment:
how about:
* `get_slots_since(counter: u64) -> Vec<(u64,u64)>` - returns new slots since counter as a vector of tuples (base_slot_index: u64, next_slot_index: u64)
* `get_slot(slot_index: u64) -> Vec<Entry>` - returns the entry vector for the slot

Reply:
That works beautifully! Some points of clarification:

1) Should SlotData.next be a list of potential next slots instead of a single u64?
2) Is it correct to assume get_slots_since() and get_slot_entries() are no longer subscription-based systems that alert Replay stage when entries of interest are available? So it will be up to ReplayStage to poll the db_ledger with these two functions to fetch updates?

Reply:
1. Either would work. Replay stage needs to maintain a graph of checkpoints anyways
2. There should still be a signal from the ledger that tells the replay stage that there is a change. The signal should fire when there are additional consecutive entries to process
3.  Since the entries returned start from block 0, replay stage shouldn’t need the block numbers.  The ledger did all the work of coalescing already.

Reply:
For 2, should the signal only fire based on slots that ReplayStage has subscribed to, or if there are any consecutive entries across any slot?

Reply:
2. I am suggesting we don't don't need an explicit subscribe.  The replay stage can figure out if any of the new data is useful, otherwise you need to replicate and track state in the db ledger that replay stage is already tracking.  Any time a sequence of entries starting from the first entry in block 0 is extended the ledger can notify the replay stage that there is something new.

Reply:
@anatoly, that means for every new blob added to some slot "n", you would have to know if slots "n" chained all the way down to slot 0, i.e. naively, you would have to check the path of blocks "n", "n-1", "n-2".. 0.

For instance assume we have full slot 0, slot 1, and empty slot 2. Then if blob 0 for slot 3 comes in chaining to slot 2, we have to know not to signal, but if blob 0 comes in for slot 4 chaining to slot 1, then we have to signal to ReplayStage that there are changes.

It sounds to do this efficiently, like we then also need a "is_trunk" boolean per slot that is true iff every block from 0...slot forms a full sequence without any holes. Then we can derive for new slot n+1:

is_trunk(n+1) iff (is_trunk(n) and  n.contains_all_ticks()). 

Then every time we complete slot_A and slot_A.is_trunk(), we also have to recursively go through every path deriving from slot_A.next and mark each of those slots as is_trunk().

Is this what you were imagining?

Reply:
@carllin the replay stage maintains a set of valid forks it can vote on.  So it will ask the ledger for new blobs for those forks, or a new fork that is derived from any of the valid forks.  The only state that the ledger needs to maintain is if the new blob extends a fork or not.  If it lands somewhere in the future and doesn't connect to anything, it doesn't need to signal the replay stage.

Reply:
Awesome, I have a v1 implementation of that here: https://github.com/solana-labs/solana/pull/2277/files.